File: python3/pr032.py
#! /usr/bin/env python3

"""
Problem 32

06 December 2002


We shall say that an n-digit number is pandigital if it makes use of all
the digits 1 to n exactly once; for example, the 5-digit number, 15234, is
1 through 5 pandigital.

The product 7254 is unusual, as the identity, 39 × 186 = 7254, containing
multiplicand, multiplier, and product is 1 through 9 pandigital.

Find the sum of all products whose multiplicand/multiplier/product
identity can be written as a 1 through 9 pandigital.

HINT: Some products can be obtained in more than one way so be sure to
only include it once in your sum.

45228
"""

from __future__ import print_function

from itertools import permutations

def main():
    result = set()
    for digits in permutations('123456789'):
        make_groupings(digits, result)

    print(sum(result))

def make_groupings(digits, result):
    digits = ''.join(digits)
    def piece(a, b):
        return int(digits[a:b])
    size = len(digits)
    for i in range(1, size-2):
        # Turns out that the only meaningful position for j is 5.
        for j in range(i+1, size-1):
            a = piece(0, i)
            b = piece(i, j)
            c = piece(j, size)
            if a*b == c:
                # print(a, b, c)
                result.add(c)

if __name__ == '__main__':
    main()

File: julia/pr032.jl
# Problem 32
#
# 06 December 2002
#
#
# We shall say that an n-digit number is pandigital if it makes use of all
# the digits 1 to n exactly once; for example, the 5-digit number, 15234, is
# 1 through 5 pandigital.
#
# The product 7254 is unusual, as the identity, 39 × 186 = 7254, containing
# multiplicand, multiplier, and product is 1 through 9 pandigital.
#
# Find the sum of all products whose multiplicand/multiplier/product
# identity can be written as a 1 through 9 pandigital.
#
# HINT: Some products can be obtained in more than one way so be sure to
# only include it once in your sum.
#
# 45228

using Combinatorics

# Extract digits out of the number.
function ofdigits(digits)
   result = 0
   for dig in digits
      result = result * 10 + dig
   end
   result
end

function make_groupings!(rset, digits)
   for i = 2:lastindex(digits)-2
      for j = i+1:lastindex(digits)-1
         a = ofdigits(digits[1:i-1])
         b = ofdigits(digits[i:j-1])
         c = ofdigits(digits[j:end])
         if a*b == c
            push!(rset, c)
         end
      end
   end
end

function solve()
   rset = Set{Int64}()
   for perm in permutations([Int8(i) for i in 1:9])
      make_groupings!(rset, perm)
   end
   sum(rset)
end

println(solve())
